test/e2e/reporters/progress.go

```diff
@@ -0,0 +1,159 @@
+limitations under the License.
+*/
+
+package reporters
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/onsi/ginkgo/config"
+	"github.com/onsi/ginkgo/types"
+	"k8s.io/klog"
+)
+
+// ProgressReporter is a ginkgo reporter which tracks the total number of tests to be run/passed/failed/skipped.
+// As new tests are completed it updates the values and prints them to stdout and optionally, sends the updates
+// to the configured URL.
+type ProgressReporter struct {
+	NumToBeRun   int      `json:"numToBeRun"`
+	NumCompleted int      `json:"numCompleted"`
+	NumSkipped   int      `json:"numSkipped"`
+	Failures     []string `json:"failures,omitempty"`
+
+	progressURL string
+	client      *http.Client
+}
+
+type progressMsg struct {
+	Msg          string   `json:"msg"`
+	NumToBeRun   int      `json:"total"`
+	NumCompleted int      `json:"completed"`
+	NumFailures  int      `json:"numFailures"`
+	NumSkipped   int      `json:"numSkipped"`
+	Failures     []string `json:"failures,omitempty"`
+}
+
+// NewProgressReporter returns a progress reporter which posts updates to the given URL.
+func NewProgressReporter(progressUpdateURL string) *ProgressReporter {
+	c := &http.Client{
+		Timeout: time.Second * 10,
+	}
+	return &ProgressReporter{
+		Failures:    []string{},
+		progressURL: progressUpdateURL,
+		client:      c,
+	}
+}
+
+// SpecSuiteWillBegin is invoked by ginkgo when the suite is about to start and is the first point in which we can
+// antipate the number of tests which will be run.
+func (reporter *ProgressReporter) SpecSuiteWillBegin(cfg config.GinkgoConfigType, summary *types.SuiteSummary) {
+	reporter.NumToBeRun = summary.NumberOfSpecsThatWillBeRun
+	reporter.sendUpdates("Test Suite starting")
+}
+
+// SpecSuiteDidEnd is the last method invoked by Ginkgo after all the specs are run.
+func (reporter *ProgressReporter) SpecSuiteDidEnd(summary *types.SuiteSummary) {
+	reporter.sendUpdates("Test Suite completed")
+}
+
+// SpecDidComplete is invoked by Ginkgo each time a spec is completed (including skipped specs).
+func (reporter *ProgressReporter) SpecDidComplete(specSummary *types.SpecSummary) {
+	testname := strings.Join(specSummary.ComponentTexts[1:], " ")
+	msg := ""
+	switch specSummary.State {
+	case types.SpecStateFailed:
+		if len(specSummary.ComponentTexts) > 0 {
+			reporter.Failures = append(reporter.Failures, testname)
+		}
+		msg = fmt.Sprintf("FAILED %v", testname)
+	case types.SpecStatePassed:
+		reporter.NumCompleted++
+		msg = fmt.Sprintf("PASSED %v", testname)
+	case types.SpecStateSkipped:
+		reporter.NumSkipped++
+		return
+	default:
+		return
+	}
+
+	reporter.sendUpdates(msg)
+}
+
+// sendUpdates serializes the current progress and prints it to stdout and also posts it to the configured endpoint if set.
+func (reporter *ProgressReporter) sendUpdates(msg string) {
+	b := reporter.serializeWithMsg(msg, true)
```

Comment:
As the other comment mentions, I opted to default to print the whole list of failures each time. This helps if you have just one of those values to parse and don't want to have to dig through all the log history to catch the list of the other failures.